smart/plugin.go

```diff
@@ -97,13 +116,58 @@ func validateName(namespace []string) bool {
 	return true
 }
 
+// Function to check properness of configuration parameters
+// and set plugin attribute accordingly
+func (sc *SmartCollector) setProcDevPath(cfg interface{}) error {
+	sc.initializedMutex.Lock()
+	defer sc.initializedMutex.Unlock()
+	procPath, err := config.GetConfigItem(cfg, "proc_path")
+	if err == nil && len(procPath.(string)) > 0 {
+		procPathStats, err := os.Stat(procPath.(string))
+		if err != nil {
+			return err
+		}
+		if !procPathStats.IsDir() {
+			return errors.New(fmt.Sprintf("%s is not a directory", procPath.(string)))
+		}
+		sc.proc_path = procPath.(string)
+	}
+	devPath, err := config.GetConfigItem(cfg, "dev_path")
+	if err == nil && len(devPath.(string)) > 0 {
+		devPathStats, err := os.Stat(devPath.(string))
+		if err != nil {
+			return err
+		}
+		if !devPathStats.IsDir() {
+			return errors.New(fmt.Sprintf("%s is not a directory", devPath.(string)))
+		}
+		sc.dev_path = devPath.(string)
+	}
+	if sysUtilProvider == nil {
+		sysUtilProvider = NewSysutilProvider(sc.proc_path, sc.dev_path)
+	}
+	sc.initialized = true
+	return nil
+}
+
 type SmartCollector struct {
+	initialized      bool
+	initializedMutex *sync.Mutex
+	logger           *log.Logger
+	proc_path        string
+	dev_path         string
 }
 
 type smartResults map[string]interface{}
 
 // CollectMetrics returns metrics from smart
 func (sc *SmartCollector) CollectMetrics(mts []plugin.MetricType) ([]plugin.MetricType, error) {
+	if !sc.initialized {
```

Comment:
You should have that check done also under Lock. Having lock just during initialization prevents data corruption during initialization (like concurrent write to complex data structure like `map`), but does not prevent double initialization (that might lead to incorrect initialization, but hopefully not in this case). This line might be kept if also added in `setProcDevPath`. Then you'll have double checked lock pattern (https://en.wikipedia.org/wiki/Double-checked_locking  ). Sorry for being that strict, but:
1. Even if now it's not causing problem it might do in future
2. Snippets from already existing plugins are copied to new ones, so we should use best known practice for such cases